### PR TITLE
Secure the bag

### PR DESCRIPTION
**Pull request by freddiecoleman**
This PR introduces “Secure the bag” to the CAT admin tool.

### Secure the bag

Secure the bag takes the concept of an airdrop and flips it on its head.

Instead of the asset issuer being responsible for distributing the coins, a tree of coins is created where all of the create coin conditions are hardcoded into the puzzles of each node.

No signature is required to spend these coins which means that anybody can spend them. This opens up some options; the asset issuer could still spend the coins and perform the airdrop themselves, but it’s also possible for the recipient to claim their own coins themselves. It’s also possible for a friendly third party who is willing to pay the fees for some reason to spend the coins.

A coin announcement of b"$" is hardcoded into the puzzle. This means that the spender can spend an XCH coin for fees which asserts this announcement.

Secure the bag also supports CATs.

### Unwrapping the secured bag

The spender only needs to spend any unspent coins in the path towards the coins that they want to unwrap and not the entire tree.

This implementation provides a parent puzzle lookup table where a spender can look up the coin id and puzzle reveal of the parent of the coin the they want to create. This allows them to spend that coin if it has been created but not spent. If the coin has not been created yet the lookup table can be used again to find the parent of that coin which can be repeated recursively until you reach the root of the tree.

### Duplicate puzzle hash + amount

This secure the bag implementation does not currently support multiple targets with the same puzzle hash and amount. It is possible to support this in future however as it introduces some complexity it was left out of the initial version.

Target data should be validated to ensure no duplicate puzzle hash + amount combination before being passed into this function.

## How to use

First use the cats command to select a coin;

```
cats --tail ./reference_tails/genesis_by_coin_id.clsp.hex --send-to txch1u0xpykzvx3nfuwnhrqvruvdj9jx6ammf3ft7nkhzhk5z7ge7u3hqvjcdp2 --amount 1000000000000 --as-bytes --select-coin                                                  
```

This has an output like;

```
{
    "amount": 88662871640000,
    "parent_coin_info": "0x56c9bde486ee4cb09f2566a08d40445dd93fcda6cc3719f5e8aeab1a7357d087",
    "puzzle_hash": "0xf33c61dbe811685944c1b505856f94d37ebeba54dd21a53935e3e48c658a78cb"
}
Name: fba7151be6edf6364e6b98567890824bc66a1060adfcbca0f2b85daaf73a3470
```

Can then execute this to get the target puzzle hash;

```
secure_the_bag --tail ./reference_tails/genesis_by_coin_id.clsp.hex --amount 1000000000000  --secure-the-bag-targets-path ./doge.csv --curry 0xfba7151be6edf6364e6b98567890824bc66a1060adfcbca0f2b85daaf73a3470 --prefix txch
```

Which has an output such as;

```
Secure the bag root puzzle hash: e3ad7e6118e8d32132f0a89c5a25c70e8d5a11660e2876ae6f28b88205d6a717
Secure the bag root address: txch1u0xpykzvx3nfuwnhrqvruvdj9jx6ammf3ft7nkhzhk5z7ge7u3hqvjcdp2
```

Then execute this to send to secure the bag;

```
cats --tail ./reference_tails/genesis_by_coin_id.clsp.hex --send-to txch1u0xpykzvx3nfuwnhrqvruvdj9jx6ammf3ft7nkhzhk5z7ge7u3hqvjcdp2 --amount 1000000000000 --as-bytes --curry 0xfba7151be6edf6364e6b98567890824bc66a1060adfcbca0f2b85daaf73a3470 
```

This has the following output;

```
Successfully pushed the transaction to the network
Asset ID: 400e7c2a4aaf545679fb839df2dbbf17410dde63a0ed9511a3cd10aff7a5d6c9
Eve Coin ID: c20353585a9fb76b5c853eca093fd157cc62c2c439ba702cdd359a0601e8e7d0
```

Once the coin has been created you can fully unwind it with;

```
unwind_the_bag --eve-coin-id c20353585a9fb76b5c853eca093fd157cc62c2c439ba702cdd359a0601e8e7d0 --tail-hash 400e7c2a4aaf545679fb839df2dbbf17410dde63a0ed9511a3cd10aff7a5d6c9 --secure-the-bag-targets-path ./doge.csv --unwind-fee 500000 --wallet-id 1
```

Or to a specific target puzzle hash with;

```
unwind_the_bag --eve-coin-id c20353585a9fb76b5c853eca093fd157cc62c2c439ba702cdd359a0601e8e7d0 --tail-hash 400e7c2a4aaf545679fb839df2dbbf17410dde63a0ed9511a3cd10aff7a5d6c9 --secure-the-bag-targets-path ./doge.csv --unwind-fee 500000 --wallet-id 1 --unwind-target-puzzle-hash bcfa768c2f7ff2882b8093be6aa7149b8030e41235a25ff7f29cbc7a8637211a
```

### CODE REVIEW

File: cats/secure_the_bag.py [hunk 1]
```diff
@@ -0,0 +1,106 @@
...
+class Target:
+    puzzle_hash: bytes32
+    amount: uint64
+
+    def __init__(self, puzzle_hash: bytes32, amount: uint64) -> None:
+        self.puzzle_hash = puzzle_hash
+        self.amount = amount
+    
+    def create_coin_condition(self) -> Tuple[bytes, bytes32, uint64, Tuple[bytes32]]:
+        return [ConditionOpcode.CREATE_COIN, self.puzzle_hash, self.amount, [self.puzzle_hash]]
```
Comment (almogdepaz) [inline cats/secure_the_bag.py:64]: it looks like the types got mixed up here, this returns a list but should return a Tuple.

what is the reason for that last Tuple[Byte32]  ?

File: cats/secure_the_bag.py [hunk 2]
```diff
@@ -0,0 +1,106 @@
...
+        current_batch.append(target)
+        
+        if len(current_batch) == leaf_width or index == len(targets) - 1:
+            results.append(current_batch)
+            current_batch = []
+
+    return results
+
+
+def secure_the_bag(targets: List[Target], leaf_width: int, asset_id: Union[bytes32, None] = None, parent_puzzle_lookup: Dict[str, Coin] = {}) -> Tuple[bytes32, Dict[str, Coin]]:
```
Comment (almogdepaz) [inline cats/secure_the_bag.py]: i think avoiding the recursion would be cleaner


```suggestion
def secure_the_bag(targets: List[Target], leaf_width: int, asset_id: Union[bytes32, None] = None, parent_puzzle_lookup: Dict[str, Coin] = {}) -> Tuple[bytes32, Dict[str, Coin]]:
    """
    Calculates secure the bag root puzzle hash and provides parent puzzle reveal lookup table for spending.

    Secures bag of CATs if optional asset id is passed.
    """
    while len(targets) > 1:
        targets = handle_batches(asset_id, batch_the_bag(targets, leaf_width), parent_puzzle_lookup)

    if asset_id is not None:
        return construct_cat_puzzle(CAT_MOD, asset_id, targets[0].puzzle_hash).get_tree_hash(
            targets[0].puzzle_hash), parent_puzzle_lookup

    return targets[0].puzzle_hash, parent_puzzle_lookup


def handle_batches(asset_id, batched_targets, parent_puzzle_lookup):
    results: List[Target] = []
    for batch_targets in batched_targets:

        list_of_conditions = [EMPTY_COIN_ANNOUNCEMENT]
        total_amount = uint64(0)

        for target in batch_targets:
            list_of_conditions.append(target.create_coin_condition())
            total_amount += target.amount

        puzzle = Program.to((1, list_of_conditions))
        puzzle_hash = puzzle.get_tree_hash()
        amount = total_amount

        results.append(Target(puzzle_hash, amount))

        if asset_id is not None:
            outer_puzzle = construct_cat_puzzle(CAT_MOD, asset_id, puzzle)

        for target in batch_targets:
            if asset_id is not None:
                target_outer_puzzle_hash = construct_cat_puzzle(CAT_MOD, asset_id, target.puzzle_hash).get_tree_hash(
                    target.puzzle_hash)
                parent_puzzle_lookup[target_outer_puzzle_hash.hex()] = Coin(target, outer_puzzle, amount)
            else:
                parent_puzzle_lookup[target.puzzle_hash.hex()] = Coin(target, puzzle, amount)
    return results
```

File: cats/secure_the_bag.py (hunk 1 above)
Comment (freddiecoleman) [reply cats/secure_the_bag.py:64]: The `Tuple[bytes32]` is the hint containing the inner puzzle hash for full nodes in index.

Comment (almogdepaz) [reply cats/secure_the_bag.py:64]: i was wondering why we wrap that bytes32 inside a Tuple, i see this is what we do in chia-blockchain so not related to this pr

**Review by Quexington — CHANGES_REQUESTED** [review]
I have two changes to request:

1) The secure the bag drivers should not have any CAT logic in them.  It's useful for constructing a series of _inner_ puzzles.  If you do it as currently implemented, then the eve spend needs to be completed by all of the recipients which means a single issuance (genesis by ID) CAT is impossible and a multi issuance (delegated TAIL) CAT requires as many signatures as recipients.  Instead, the CAT should be minted in its entirety first, then given the inner puzzle hash that this tool spits out.  Whatever tool is used to unwrap the airdrop can handle calculating CAT puzzle hashes and filling out lineage proofs and such.

2) We should add some simulator tests of actually spending one of these rather than just calculating out the puzzle hashes and ensuring they match.  Maybe when the unwrapping tool is complete, we can add that as a test for it.

**Review by Quexington — APPROVED** [review]
Great work!

**Pull request merged**